### PR TITLE
Improve collided exception performance

### LINKED ISSUE #99965: [Perf] Windows/x64: 2 Regressions on 3/16/2024 1:13:28 AM
### Run Information

Name | Value
-- | --
Architecture | x64
OS | Windows 10.0.18362
Queue | TigerWindows
Baseline | [f36aa7ec736735a3e92046c73d6e03029017b9fa](https://github.com/dotnet/runtime/commit/f36aa7ec736735a3e92046c73d6e03029017b9fa)
Compare | [6f10cd3b1c7384057f8c35909fe093cbc4106ff2](https://github.com/dotnet/runtime/commit/6f10cd3b1c7384057f8c35909fe093cbc4106ff2)
Diff | [Diff](https://github.com/dotnet/runtime/compare/f36aa7ec736735a3e92046c73d6e03029017b9fa...6f10cd3b1c7384057f8c35909fe093cbc4106ff2)
Configs | CompilationMode:tiered, RunKind:micro

### Regressions in Exceptions.Handling

Benchmark | Baseline | Test | Test/Base | Test Quality | Edge Detector | Baseline IR | Compare IR | IR Ratio
-- | -- | -- | -- | -- | -- | -- | -- | --
|<ul><li>[CatchAndThrowOtherDeep - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_Windows 10.0.18362/Exceptions.Handling.CatchAndThrowOtherDeep(kind%3a%20Software).html>)</li><li>📝 - [Benchmark Source](<https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/runtime/Exceptions/Handling.cs#L418-#L433>)</li><li>[📈 - ADX Test Multi Config Graph](<https://dataexplorer.azure.com/clusters/dotnetperf.westus/databases/PerformanceData?query=H4sIAAAAAAAACo1UXW/TMBR9R+I/+C0JZG1augKVgui6wsYolC5vCCE3uU0M8YeunTWZ0H47dlqVtZNQpCiyb8459/jKJwugukLgIIx+9odsC0AgFxUrsxUoSeKYeJk0AkwfK2EYB+8YdYFUpEWLQ9jofgE0031OmTgBTjEtmIHU2G6ECeJ79XjkhcSjyO0iOKAT0OYL5eA033vzOgVlmBS6d0VFVjKR92bUpMVUZEmBcvvVWNYlgPJ/M5FNyK3cmC1FCP71X1ViJsWG5RXSVuq7N5NcsbLdLWQG3g8SPxDPMAvPvOf/Y9rSjW20Z3CWonxEYBq4Mo3/hDiVJugAu5b6c3nH24l1wX/Sc5EzAV2wTrgLbiGF7GjXQa+FAVQI9t2Fssxl0qhOhpdFo1lKyyVKLl2lK/GqWSPLPpZyTUt239a60FbDVdcO81oB2jSI9qoeX96ZtEkBTKTar46DkFjaraFckXckowZcqPxhNBydRYOzwdskejM5t8+4NzofDMbD1y+jaBJFroWuOKfI7oFQzH9yWvutoHMQkhcBWTcnHcJDmEJ3lb9VUIHVUSh/2SA+Aa9AV6WJ0brO/Ecn2dWnd4A0t0qvgpC0Uk44/oAMbC6bQ8UNq904y2i/ARJ3xrSgaMiWmYL42g4PdHzghKROZVlxEZ96anZ1He9MWKBhpoTYu7Sjs7+PZr91FOILHbiaViUzMa1BB38BNJ8c++AEAAA=>)</li></ul> | 170.23 μs | 200.51 μs | 1.18 | 0.20 | False | | |

graph
[Test Report](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_19_2024/refs/heads/main_x64_Windows%2010.0.18362_Regression/Exceptions.Handling.html>)

### Repro
General Docs link: https://github.com/dotnet/performance/blob/main/docs/benchmarking-workflow-dotnet-runtime.md

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f net8.0 --filter 'Exceptions.Handling*'
```

<details>

### Payloads

[Baseline](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-79ed3a65-0c13-4ec4-89e5-865b5269652f389a120092f4ac897/c8391358-c5b0-4628-9deb-495f89f84343.zip?sv=2021-08-06&se=2024-04-14T16%3A35%3A26Z&sr=c&sp=rl&sig=McpJxxOxw82oyZCeN0h7FrhN73fmF7rB1RM4TEhFRfg%3D>)
[Compare](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-e360ebcb-2f87-40e7-a073-337c2ee2c7eb5582cb4f6174cbab1/bc03e234-e2ad-492e-a26c-542785d262ff.zip?sv=2021-08-06&se=2024-04-15T11%3A12%3A55Z&sr=c&sp=rl&sig=zrI6Ijn8rYdIa3BCd%2BlE%2FZTE9mwECKa3iIaCf37OmDM%3D>)
### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

---

### Run Information

Name | Value
-- | --
Architecture | x64
OS | Windows 10.0.18362
Queue | TigerWindows
Baseline | [f36aa7ec736735a3e92046c73d6e03029017b9fa](https://github.com/dotnet/runtime/commit/f36aa7ec736735a3e92046c73d6e03029017b9fa)
Compare | [6f10cd3b1c7384057f8c35909fe093cbc4106ff2](https://github.com/dotnet/runtime/commit/6f10cd3b1c7384057f8c35909fe093cbc4106ff2)
Diff | [Diff](https://github.com/dotnet/runtime/compare/f36aa7ec736735a3e92046c73d6e03029017b9fa...6f10cd3b1c7384057f8c35909fe093cbc4106ff2)
Configs | CompilationMode:tiered, RunKind:micro

### Regressions in System.Collections.Tests.Perf_BitArray

Benchmark | Baseline | Test | Test/Base | Test Quality | Edge Detector | Baseline IR | Compare IR | IR Ratio
-- | -- | -- | -- | -- | -- | -- | -- | --
|<ul><li>[BitArrayBitArrayCtor - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_Windows 10.0.18362/System.Collections.Tests.Perf_BitArray.BitArrayBitArrayCtor(Size%3a%20512).html>)</li><li>📝 - [Benchmark Source](<https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/libraries/System.Collections/Perf.BitArray.cs#L36-#L38>)</li><li>[📈 - ADX Test Multi Config Graph](<https://dataexplorer.azure.com/clusters/dotnetperf.westus/databases/PerformanceData?query=H4sIAAAAAAAACo1U74/SQBD9buL/sN+21R4UjgMhqRHw1FMxCHwz5rK0A12zP5rZ7R29GP92dwuHB5eYfml3pu+9eTuZ6QyYKREkKGte/Cb3OSCQSclFtoBCkyQhNNNWgW1jqSyXQE9RE2QqzWscwsa0c2CZaUvG1RlwjGnOLaTWVSNckYDu+j0aEcpQukN4RK/A2G9Mgtd8R5eVsSBbUy2E43KtTMsDTGsOuLmdcDtGZFXr8fD4nlqNwZI/wIhcdbrhPy+LUk212vBtiayW+0GnWhZc1NFMZ0B/kuQPoZY7eEZf/o/pUl+4yg4MyVPUTwjcgCxsFTwjjrUNG8ButPkq7mTdvSb4z+ZabbmCJlgv3AQ300o3tOuhN8oCFgju2YQy3+pVVTQyPM8rw1Mm5qil9pmmxE/VGnn2Ueg1E/yhzjWhLbqLphWudwWg2wxVj+3pIE+12xrAlS4Op9OlWDna0jJZkLckYxb8ggXduNu7iDsXneEqfjO6Go56/VZ3MBheDvqv43gUx76EKaVk6AacMNzeSrYLakHvICKvQrKuzipEx8WK/Ch/L6EEp1Og/uUW6xl4AaYUNkHnOgue3GSfH98Bsq1TugwjUkt54eQDclCZqI4Z36w68JbRfQMk/o5pztCSe25zEhjXPDDJkRORXapFKVVy7qna502yN+GAllsBCX3vWud+JdUh9BQSKBP6nCkEtwnbgQn/Astt9IjsBAAA>)</li></ul> | 18.73 ns | 20.97 ns | 1.12 | 0.14 | False | | |

graph
[Test Report](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_19_2024/refs/heads/main_x64_Windows%2010.0.18362_Regression/System.Collections.Tests.Perf_BitArray.html>)

### Repro
General Docs link: https://github.com/dotnet/performance/blob/main/docs/benchmarking-workflow-dotnet-runtime.md

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f net8.0 --filter 'System.Collections.Tests.Perf_BitArray*'
```

<details>

### Payloads

[Baseline](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-79ed3a65-0c13-4ec4-89e5-865b5269652f389a120092f4ac897/c8391358-c5b0-4628-9deb-495f89f84343.zip?sv=2021-08-06&se=2024-04-14T16%3A35%3A26Z&sr=c&sp=rl&sig=McpJxxOxw82oyZCeN0h7FrhN73fmF7rB1RM4TEhFRfg%3D>)
[Compare](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-e360ebcb-2f87-40e7-a073-337c2ee2c7eb5582cb4f6174cbab1/bc03e234-e2ad-492e-a26c-542785d262ff.zip?sv=2021-08-06&se=2024-04-15T11%3A12%3A55Z&sr=c&sp=rl&sig=zrI6Ijn8rYdIa3BCd%2BlE%2FZTE9mwECKa3iIaCf37OmDM%3D>)
### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### PR DESCRIPTION
**Pull request by janvorli**
With the new EH enabled, one exception handling performance test has regressed while all other improved dramatically. I have investigated the test case and it turned out that the regression is due to the way we unwind during second pass when we have an exception that occured in a catch or finally funclet call chain and escaped it. What we do is that we unwind stack until we reach the parent stack frame of the catch / finally and then continue searching for handlers. The NativeAOT that the new EH is based on doesn't unwind stack though, it just moves the current stack frame iterator to the position of the previous exception's stack frame iterator by copying its state. I have applied the same mechanism to the new EH in coreclr and it improved the performance of that test 3-4 times on my machine.

Close #99965

### CODE REVIEW

**Comment by janvorli** [comment]
The arm gc stress issue is known - see #99410

File: src/coreclr/vm/stackwalk.cpp [hunk 1]
```diff
@@ -1546,6 +1546,70 @@ BOOL StackFrameIterator::IsValid(void)
...
+// Advance to the position that the other iterator is currently at.
+//
+void StackFrameIterator::SkipTo(StackFrameIterator *pOtherStackFrameIterator)
+{
+    // We copy the other stack frame iterator over the current one, but we need to
+    // keep a couple of members untouched. So we save them here and restore them
+    // after the copy.
+    ExInfo* pPrevExInfo = GetNextExInfo();
+    REGDISPLAY *pRD = m_crawl.GetRegisterSet();
+    GSCookie *pCurGSCookie = m_crawl.pCurGSCookie;
```
Comment (jkotas) [inline src/coreclr/vm/stackwalk.cpp:1562]: Why do we need to patch GS cookie pointers?

I would expect that the GS cookie pointer moves in a lock step with registers, and so the one from `pOtherStackFrameIterator` should work fine.

**Review by jkotas — APPROVED** [review]

File: src/coreclr/vm/stackwalk.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/vm/stackwalk.cpp:1562]: Maybe we don't, I just made it equivalent to what the loop that was there before was getting to. I'll check if we can keep the ones we've copied instead.

Comment (janvorli) [reply src/coreclr/vm/stackwalk.cpp]: You were right, it works fine without that too. I've added a commit to remove it.

**Pull request merged**